File: BorderlessWindow/Border.py
from enum import Enum

from PySide6.QtCore import Qt
from PySide6.QtGui import QMouseEvent, QPaintEvent, QPainter, QColor
from PySide6.QtWidgets import QWidget


class BorderType(Enum):
    top_border = 1  # 上
    bottom_border = 2  # 下
    left_border = 3  # 左
    right_border = 4  # 右
    left_top_border = 5  # 左上
    right_top_border = 6  # 右上
    left_bottom_border = 7  # 左下
    right_bottom_border = 8  # 右下


class Border(QWidget):
    def __init__(self, parent: QWidget, border_type: BorderType):
        super().__init__(parent)
        self.setMouseTracking(True)  # 鼠标追踪

        self.border_type = border_type  # 边框类型
        self.border_width = 5  # 边框宽度
        self.FixedSize()

    def mousePressEvent(self, event: QMouseEvent):
        if self.window().isMaximized():
            return super().mousePressEvent(event)
        if event.button() == Qt.LeftButton:
            match self.border_type:
                case BorderType.top_border:  # 上
                    self.window().windowHandle().startSystemResize(Qt.Edge.TopEdge)
                case BorderType.bottom_border:  # 下
                    self.window().windowHandle().startSystemResize(Qt.Edge.BottomEdge)
                case BorderType.left_border:  # 左
                    self.window().windowHandle().startSystemResize(Qt.Edge.LeftEdge)
                case BorderType.right_border:  # 右
                    self.window().windowHandle().startSystemResize(Qt.Edge.RightEdge)
                case BorderType.left_top_border:  # 左上
                    self.window().windowHandle().startSystemResize(Qt.Edge.LeftEdge | Qt.Edge.TopEdge)
                case BorderType.right_bottom_border:  # 右下
                    self.window().windowHandle().startSystemResize(Qt.Edge.RightEdge | Qt.Edge.BottomEdge)
                case BorderType.left_bottom_border:  # 左下
                    self.window().windowHandle().startSystemResize(Qt.Edge.LeftEdge | Qt.Edge.BottomEdge)
                case BorderType.right_top_border:  # 右上
                    self.window().windowHandle().startSystemResize(Qt.Edge.RightEdge | Qt.Edge.TopEdge)
        return super().mousePressEvent(event)

    def mouseMoveEvent(self, event: QMouseEvent):
        if self.window().isMaximized():
            self.setCursor(Qt.CursorShape.ArrowCursor)
            return super().mouseMoveEvent(event)
        match self.border_type:
            case BorderType.top_border | BorderType.bottom_border:  # 上下
                self.setCursor(Qt.CursorShape.SizeVerCursor)
            case BorderType.left_border | BorderType.right_border:  # 左右
                self.setCursor(Qt.CursorShape.SizeHorCursor)
            case BorderType.left_top_border | BorderType.right_bottom_border:  # 左上右下
                self.setCursor(Qt.CursorShape.SizeFDiagCursor)
            case BorderType.left_bottom_border | BorderType.right_top_border:  # 左下右上
                self.setCursor(Qt.CursorShape.SizeBDiagCursor)
        return super().mouseMoveEvent(event)

    def paintEvent(self, event: QPaintEvent):
        painter = QPainter(self)
        painter.setPen(Qt.PenStyle.NoPen)
        painter.setBrush(QColor(255, 255, 255, 1))
        painter.drawRect(self.rect())

    def FixedSize(self):
        """固定大小"""
        if (self.border_type == BorderType.top_border or
                self.border_type == BorderType.bottom_border):
            return self.setFixedHeight(self.border_width)
        elif (self.border_type == BorderType.left_border or
              self.border_type == BorderType.right_border):
            return self.setFixedWidth(self.border_width)
        else:
            return self.setFixedSize(self.border_width, self.border_width)

File: Archive scrap code/BorderlessWindow/border.cpp
#include "border.h"

Border::Border(QWidget *parent, BorderType bordertype)
    : QWidget{parent},
      border_type{bordertype}
{
    this->setMouseTracking(true);

    if (this->border_type == BorderType::top_border || this->border_type == BorderType::bottom_border)
        this->setFixedHeight(this->border_width);
    else if (this->border_type == BorderType::left_border || this->border_type == BorderType::right_border)
        this->setFixedWidth(this->border_width);
    else
        this->setFixedSize(this->border_width, this->border_width);
}

void Border::mousePressEvent(QMouseEvent *event)
{
    if (this->window()->isMaximized())
        return QWidget::mousePressEvent(event);
    if (event->button() == Qt::LeftButton)
    {
        switch (this->border_type)
        {
        case BorderType::top_border:
            this->window()->windowHandle()->startSystemResize(Qt::Edge::TopEdge);
            break;
        case BorderType::bottom_border:
            this->window()->windowHandle()->startSystemResize(Qt::Edge::BottomEdge);
            break;
        case BorderType::left_border:
            this->window()->windowHandle()->startSystemResize(Qt::Edge::LeftEdge);
            break;
        case BorderType::right_border:
            this->window()->windowHandle()->startSystemResize(Qt::Edge::RightEdge);
            break;
        case BorderType::left_top_border:
            this->window()->windowHandle()->startSystemResize(Qt::Edge::LeftEdge | Qt::Edge::TopEdge);
            break;
        case BorderType::right_bottom_border:
            this->window()->windowHandle()->startSystemResize(Qt::Edge::RightEdge | Qt::Edge::BottomEdge);
            break;
        case BorderType::left_bottom_border:
            this->window()->windowHandle()->startSystemResize(Qt::Edge::LeftEdge | Qt::Edge::BottomEdge);
            break;
        case BorderType::right_top_border:
            this->window()->windowHandle()->startSystemResize(Qt::Edge::RightEdge | Qt::Edge::TopEdge);
            break;
        }
    }
    return QWidget::mousePressEvent(event);
}

void Border::mouseMoveEvent(QMouseEvent *event)
{
    if (this->window()->isMaximized())
    {
        this->setCursor(Qt::CursorShape::ArrowCursor);
        return QWidget::mouseMoveEvent(event);
    }
    switch (this->border_type)
    {
    case BorderType::top_border:
        this->setCursor(Qt::CursorShape::SizeVerCursor);
        break;
    case BorderType::bottom_border:
        this->setCursor(Qt::CursorShape::SizeVerCursor);
        break;
    case BorderType::left_border:
        this->setCursor(Qt::CursorShape::SizeHorCursor);
        break;
    case BorderType::right_border:
        this->setCursor(Qt::CursorShape::SizeHorCursor);
        break;
    case BorderType::left_top_border:
        this->setCursor(Qt::CursorShape::SizeFDiagCursor);
        break;
    case BorderType::right_bottom_border:
        this->setCursor(Qt::CursorShape::SizeFDiagCursor);
        break;
    case BorderType::left_bottom_border:
        this->setCursor(Qt::CursorShape::SizeBDiagCursor);
        break;
    case BorderType::right_top_border:
        this->setCursor(Qt::CursorShape::SizeBDiagCursor);
        break;
    }
    return QWidget::mouseMoveEvent(event);
}

void Border::paintEvent(QPaintEvent *)
{
    QPainter painter(this);
    painter.setPen(Qt::PenStyle::NoPen);
    painter.setBrush(QColor(0,0,0,1));
    painter.drawRect(rect());
}
